src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/SpecimenMapper.java

```diff
@@ -1,14 +1,87 @@
 package uk.nhs.digital.nhsconnect.lab.results.translator.mapper;
 
+import lombok.RequiredArgsConstructor;
+import org.hl7.fhir.dstu3.model.Annotation;
+import org.hl7.fhir.dstu3.model.Identifier;
 import org.hl7.fhir.dstu3.model.Specimen;
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Component;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.FreeTextSegment;
 import uk.nhs.digital.nhsconnect.lab.results.model.edifact.Message;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.Reference;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.SpecimenCharacteristicType;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.SpecimenQuantity;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.segmentgroup.SpecimenDetails;
+import uk.nhs.digital.nhsconnect.lab.results.utils.UUIDGenerator;
 
+import java.util.Arrays;
 import java.util.List;
+import java.util.Optional;
+
+import static java.util.stream.Collectors.toList;
 
 @Component
+@RequiredArgsConstructor(onConstructor_ = @Autowired)
 public class SpecimenMapper {
+    private static final String IDENTIFIER_SYSTEM = "http://ehr.acme.org/identifiers/collections";
+    private static final String ACCESSION_IDENTIFIER_SYSTEM = "http://lab.acme.org/specimens/2011";
+    private static final String TYPE_SYSTEM = "http://snomed.info/sct";
+
+    private final UUIDGenerator uuidGenerator;
+    private final DateFormatMapper dateFormatMapper;
+
     public List<Specimen> mapToSpecimens(final Message message) {
-        return List.of(new Specimen());
+        return message.getServiceReportDetails().getSubject().getSpecimens().stream()
+            .map(this::edifactToFhir)
+            .collect(toList());
+    }
+
+    private Specimen edifactToFhir(final SpecimenDetails edifact) {
+        final Specimen fhir = new Specimen();
```

Comment:
According to the fhir specification, the `subject` on the `Specimen` resource is required: http://hl7.org/fhir/stu3/Specimen.html

Reply:
Well spotted. I guess I missed it because I was focussed on those with edifact mappings in the spreadsheet.

I guess that the data we need is the `Reference::getNumber` of the `InvestigationSubject::getReferenceServiceSubject` (`RFF+SSI`) - what do you think?

Reply:
In terms of FHIR, the `subject` should be the id of the `Patient` resource from my understanding as it's a Reference type according to:
http://hl7.org/fhir/stu3/specimen-definitions.html#Specimen.subject

Reply:
From what I can see in this example: https://developer.nhs.uk/apis/itk3nationalpathology-1-1-0/design_singletest_bundle.html, the subject reference is the uuid generated and set on the Patient resource

Reply:
`RFF+SSI` may be the NHS number

Reply:
@saileshp-kainos - just seen your comment before mine :D

Reply:
Yep, you're both right. This complicates matters... We only set the reference in the `BundleMapper`, after the `PathologyRecord` has already been compiled from all of the edifact to fhir mappers. We either need to redesign the mappers quite drastically so that the `SpecimenMapper` can already access the `fullUrl` of the patient resource, or we amend the specimen resources in the `BundleMapper` before adding them as entries in the bundle. The latter is easiest, but seems like we'd be confusing responsibilities.

Reply:
Would this work?
<img width="705" alt="Screenshot 2021-03-01 at 10 59 35" src="https://user-images.githubusercontent.com/16392604/109488302-490cac00-7a7d-11eb-8f56-43f0f3cedcff.png">

Reply:
Hm, not according to the example you shared. In the example the reference is the fullUrl (e.g. `urn:uuid:8d6c2cd5-0eec-496a-88d0-3785a135df09`). In our project it's just `urn:uuid:` + `patient.getId()` so we could just do that, but eww... hacky!

Reply:
yep, that is true!

Reply:
i wonder how many more example we have like this - where id's/data are shared across resources

Reply:
I think what we can do here is to put the responsibility for creating `fullUrl`s of resources into a separate component that both the `BundleMapper` and the individual edifact to fhir mappers can use.

That way, for what we need in this story:
1. a Patient is mapped and given an `Id` (some-uuid)
2. the patient is passed to the `SpecimenMapper` which uses the shared component to work out that the reference to the patient _will be_ `urn:uuid:some-uuid`
3. the `BundleMapper` uses the same shared component to set the Patient's `fullUrl` to `urn:uuid:some-uuid`.

Unit tests will verify the behaviour of the shared component to always map `resource.getId()` to `"urn:uuid:" + resource.getId()` and the component will be stateless and the method idempotent.

Sound okay?

Reply:
Sounds like a good plan 👍

Reply:
Now added